### PR TITLE
Add new audio objects for loading sounds

### PR DESCRIPTION
**Pull request by IntelOrca**
All references to CSS are now moved to an object file which is loaded at the start of the game. The audio object type is a new kind of object which is not saved to park files.

New audio objects are available in this PR:
https://github.com/OpenRCT2/objects/pull/185

**They include:**
rct2.audio.base - Base audio for OpenRCT2, includes all sound effects and crowd ambience.
rct2.audio.title - The RCT2 title music.
rct2.audio.circus - The "ride music" that is played by the circus ride.
rct1.audio.title - The RCT1 title music.

The audio code is a bit of a mess. It went through a few refactors and still remains unfinished. So at some point I need to go back and clean it up.

### CODE REVIEW

**Comment by Gymnasiast** [comment]
I don’t have a lot to remark about the design, it seems fine to me. Perhaps Duncan can say more about that.

I do think it might be worth splitting off (some of?) the object refactors so everything can be tested properly.

**Comment by Broxzier** [comment]
Requires https://github.com/OpenRCT2/objects/pull/185

File: src/openrct2-ui/audio/AudioContext.cpp [hunk 1]
```diff
@@ -65,14 +67,42 @@ namespace OpenRCT2::Audio
...
+
         IAudioSource* CreateStreamFromWAV(const std::string& path) override
         {
-            return AudioSource::CreateStreamFromWAV(path);
+            return AddSource(AudioSource::CreateStreamFromWAV(path));
+        }
+
+        IAudioSource* CreateStreamFromCSS(std::unique_ptr<IStream> stream, uint32_t index) override
+        {
+            auto rw = StreamToSDL2(std::move(stream));
```
Comment (frutiemax) [inline src/openrct2-ui/audio/AudioContext.cpp:83]: auto* rw

File: src/openrct2-ui/audio/AudioContext.cpp [hunk 2]
```diff
@@ -65,14 +67,42 @@ namespace OpenRCT2::Audio
...
+            auto& format = _audioMixer->GetFormat();
+            return AddSource(AudioSource::CreateMemoryFromCSS1(rw, index, &format));
         }
 
         IAudioSource* CreateStreamFromWAV(std::unique_ptr<IStream> stream) override
         {
-            return AudioSource::CreateStreamFromWAV(std::move(stream));
+            constexpr size_t STREAM_MIN_SIZE = 2 * 1024 * 1024; // 2 MiB
+            auto loadIntoRAM = stream->GetLength() < STREAM_MIN_SIZE;
+            auto rw = StreamToSDL2(std::move(stream));
```
Comment (frutiemax) [inline src/openrct2-ui/audio/AudioContext.cpp:97]: auto* rw

File: src/openrct2-ui/audio/AudioContext.cpp [hunk 3]
```diff
@@ -120,6 +150,44 @@ namespace OpenRCT2::Audio
...
+
+    private:
+        IAudioSource* AddSource(std::unique_ptr<ISDLAudioSource> source)
+        {
+            return _audioMixer->AddSource(std::move(source));
+        }
+
+        static SDL_RWops* StreamToSDL2(std::unique_ptr<IStream> stream)
+        {
+            auto rw = SDL_AllocRW();
```
Comment (frutiemax) [inline src/openrct2-ui/audio/AudioContext.cpp:162]: auto* rw

File: src/openrct2-ui/audio/AudioContext.cpp [hunk 4]
```diff
@@ -120,6 +150,44 @@ namespace OpenRCT2::Audio
...
+            rw->read = [](SDL_RWops* ctx, void* buf, size_t size, size_t maxnum) {
+                auto ptr = static_cast<IStream*>(ctx->hidden.unknown.data1);
+                return static_cast<size_t>(ptr->TryRead(buf, size * maxnum) / size);
+            };
+            rw->size = [](SDL_RWops* ctx) {
+                auto ptr = static_cast<IStream*>(ctx->hidden.unknown.data1);
+                return static_cast<Sint64>(ptr->GetLength());
+            };
+            rw->close = [](SDL_RWops* ctx) {
+                auto ptr = static_cast<IStream*>(ctx->hidden.unknown.data1);
```
Comment (frutiemax) [inline src/openrct2-ui/audio/AudioContext.cpp:183]: auto* ptr

File: src/openrct2-ui/audio/AudioMixer.cpp [hunk 5]
```diff
@@ -1,516 +1,439 @@
...
+{
+    Close();
+
+    SDL_AudioSpec want = {};
+    want.freq = 22050;
+    want.format = AUDIO_S16SYS;
+    want.channels = 2;
+    want.samples = 2048;
+    want.callback = [](void* arg, uint8_t* dst, int32_t length) -> void {
+        auto mixer = static_cast<AudioMixer*>(arg);
```
Comment (frutiemax) [inline src/openrct2-ui/audio/AudioMixer.cpp:41]: auto* mixer

File: src/openrct2-ui/audio/AudioMixer.cpp [hunk 6]
```diff
@@ -1,516 +1,439 @@
...
-        void Stop(IAudioChannel* channel) override
-        {
-            Lock();
-            channel->SetStopping(true);
-            Unlock();
-        }
+IAudioChannel* AudioMixer::Play(IAudioSource* source, int32_t loop, bool deleteondone)
+{
+    Lock();
+    ISDLAudioChannel* channel = AudioChannel::Create();
```
Comment (frutiemax) [inline src/openrct2-ui/audio/AudioMixer.cpp:91]: auto* channel

File: src/openrct2/audio/Audio.cpp [hunk 7]
```diff
@@ -155,6 +88,14 @@ namespace OpenRCT2::Audio
                 }
             }
         }
+
+        auto& objManager = GetContext()->GetObjectManager();
+        auto baseAudio = objManager.LoadObject(AudioObjectIdentifiers::Rct2Base);
```
Comment (frutiemax) [inline src/openrct2/audio/Audio.cpp:93]: auto* baseAudio

File: src/openrct2/audio/Audio.cpp [hunk 8]
```diff
@@ -236,11 +167,16 @@ namespace OpenRCT2::Audio
         return params;
     }
 
-    void Play(SoundId soundId, int32_t volume, int32_t pan)
+    AudioObject* GetBaseAudioObject()
     {
-        if (gGameSoundsOff)
-            return;
+        auto& objManager = GetContext()->GetObjectManager();
+        auto baseAudioObject = static_cast<AudioObject*>(
```
Comment (frutiemax) [inline src/openrct2/audio/Audio.cpp:173]: auto* baseAudioObject

File: src/openrct2/audio/Audio.cpp [hunk 9]
```diff
@@ -249,7 +185,62 @@ namespace OpenRCT2::Audio
...
+        Mixer_Play_Effect(audioSource, MIXER_LOOP_NONE, DStoMixerVolume(volume), DStoMixerPan(mixerPan), 1, 1);
+    }
+
+    void Play3D(SoundId soundId, const CoordsXYZ& loc)
+    {
+        if (!IsAvailable())
+            return;
+
+        // Get sound from base object
+        auto baseAudioObject = GetBaseAudioObject();
```
Comment (frutiemax) [inline src/openrct2/audio/Audio.cpp:197]: auto* baseAudioObject

File: src/openrct2/audio/Audio.cpp [hunk 10]
```diff
@@ -249,7 +185,62 @@ namespace OpenRCT2::Audio
...
+        }
+    }
+
+    void Play(SoundId soundId, int32_t volume, int32_t pan)
+    {
+        if (!IsAvailable())
+            return;
+
+        // Get sound from base object
+        auto baseAudioObject = GetBaseAudioObject();
```
Comment (frutiemax) [inline src/openrct2/audio/Audio.cpp:218]: auto* baseAudioObject

File: src/openrct2/audio/Audio.cpp [hunk 11]
```diff
@@ -265,26 +256,24 @@ namespace OpenRCT2::Audio
             return;
         }
 
-        int32_t pathId;
-        switch (gConfigSound.title_music)
+        // Load title sequence audio object
+        auto descriptor = GetTitleMusicDescriptor();
+        auto& objManager = GetContext()->GetObjectManager();
+        auto audioObject = static_cast<AudioObject*>(objManager.LoadObject(descriptor));
```
Comment (frutiemax) [inline src/openrct2/audio/Audio.cpp:262]: auto* audioObject

File: src/openrct2/audio/Audio.cpp [hunk 12]
```diff
@@ -324,6 +313,18 @@ namespace OpenRCT2::Audio
             Mixer_Stop_Channel(gTitleMusicChannel);
             gTitleMusicChannel = nullptr;
         }
+
+        // Unload the audio object
+        if (_titleAudioObjectEntryIndex != OBJECT_ENTRY_INDEX_NULL)
+        {
+            auto& objManager = GetContext()->GetObjectManager();
+            auto obj = objManager.GetLoadedObject(ObjectType::Audio, _titleAudioObjectEntryIndex);
```
Comment (frutiemax) [inline src/openrct2/audio/Audio.cpp:321]: auto* obj

**Review by frutiemax — COMMENTED** [review]
Some auto* places to correct

File: src/openrct2/object/AudioSampleTable.h [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+#include "../core/JsonFwd.hpp"
+
+struct IReadObjectContext;
+
+class AudioSampleTable
+{
+private:
+    struct Entry
+    {
+        OpenRCT2::Audio::IAudioSource* Source{};
```
Comment (duncanspumpkin) [inline src/openrct2/object/AudioSampleTable.h:22]: Why isn't this a smart pointer.

**Comment by duncanspumpkin** [comment]
There is far too much happening in this PR to be able to review it properly. If it all works i've got nothing much to say about it the audio code as you know is not the prettiest to use.

**Review by duncanspumpkin — APPROVED** [review]

File: src/openrct2/object/AudioSampleTable.h (hunk 13 above)
Comment (IntelOrca) [reply src/openrct2/object/AudioSampleTable.h:22]: Sources are stored as smart pointers, but on the mixer.

**Pull request merged**